### PR TITLE
Basic working version with denormalizer #7

### PR DESCRIPTION
**Pull request by kevin-jueni**

### CODE REVIEW

File: src/Denner/Client/AbstractClient.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            : null;
+    }
+
+    /**
+     * @param array $result
+     * @param string $class
+     * @param null|string $rootNode
+     * @return array
+     */
+    protected function deNormalize(array $result = array(), $class, $rootNode = null)
```
Comment (curzio-della-santa) [inline src/Denner/Client/AbstractClient.php:84]: Uppercase N

File: src/Denner/Client/AbstractClient.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            )
+        );
+
+        $rootNode = null;
+
+        if (strpos($name, 'list') === 0) {
+            $rootNode = $this->getRootProperty($config['model']);
+        }
+
+        return $this->denormalize($results, $config['model'], $rootNode);
```
Comment (curzio-della-santa) [inline src/Denner/Client/AbstractClient.php:61]: Lowercase N

File: src/Denner/Client/AbstractClient.php [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * @param $model
```
Comment (curzio-della-santa) [inline src/Denner/Client/AbstractClient.php:65]: set 'string' as type

File: src/Denner/Client/ServiceDescription/Articles.php [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                    '$ref' => 'PageSizeParam',
+                ),
+                'filter' => array(
+                    '$ref' => 'FilterParam',
+                ),
+                'sort' => array(
+                    '$ref' => 'SortParam',
+                ),
+            ),
+            'model' => 'Denner\Common\Promotion\Promotion',
```
Comment (curzio-della-santa) [inline src/Denner/Client/ServiceDescription/Articles.php:66]: I would name it a bit longer, e.g. 'customModelClass'

**Comment by curzio-della-santa** [comment]
@kevin-jueni  Please fix at least the uppercase/lowercase method name, then I'll merge to be able lo use it.

Generally I think is better if you define local DTO or Models. For \Denner\Common\Promotion\Promotion is not directly  needed, but it would be for \Denner\Common\Promotion\AdvertisedArticle because is an abstract class.

File: src/Denner/Client/AbstractClient.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                sprintf(
+                    'Could not find operation for: "%s"',
+                    $name
+                )
+            );
+        }
+
+        $config = $operation->toArray();
+
+        if (!isset($config['model'])) {
```
Comment (curzio-della-santa) [inline src/Denner/Client/AbstractClient.php:39]: In my opinion should not throw  an exception, but return the array result if not present.

I would also return an array if the "\Denner\Common" Namespace is not available.

Would be much more compatible, and could be used without the require-dev of denner-common (that is a "private" repository, and not available for external consumers of the library)

EDIT: better is not to rely on "if the "\Denner\Common" Namespace is not available" but to add a configuration on the class construction (e.g.: useResultCustomModels, default false - OR - arrayResults, default true), and use accordingly

File: composer.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        {
+            "type": "vcs",
+            "url": "git@bitbucket.org:detailnet/denner-common.git"
+        }
+    ],
+    "require": {
+        "php": ">=5.5.0",
+        "guzzlehttp/guzzle": "~5.0",
+        "guzzlehttp/guzzle-services": "~0.5.0",
+        "detailnet/dfw-normalization": "~0.1.6",
```
Comment (curzio-della-santa) [inline composer.json:32]: move to require-dev and add then the suggestion "For usage of the denner model response classes" for it, 

same for "zendframework/zend-paginator" above.

**Comment by ivan-wolf** [comment]
Not implementing at the moment.

**Pull request closed without merging**